server/server.go

```diff
@@ -161,61 +230,116 @@ func (s *Server) uploadFile(uploadFile io.Reader, w http.ResponseWriter, fileNam
-			return
-		}
+	var extractorFile fileExtractor = func(r *http.Request) (uploadFile io.Reader, filename string, uerr *UserError) {
+		uploadFile, header, err := r.FormFile("image")
 
-		thumbs, err := parseThumbs(r)
 		if err != nil {
-			ErrorResponse(w, err.Error(), http.StatusBadRequest)
-			return
+			return nil, "", &UserError{LogMessage: err, UserFacingMessage: errors.New("Error processing file")}
 		}
 
-		s.uploadFile(uploadFile, w, header.Filename, thumbs)
-		uploadFile.Close()
+		return uploadFile, header.Filename, nil
 	}
 
-	urlHandler := func(w http.ResponseWriter, r *http.Request) {
-		uploadFile, err := s.download(r.FormValue("image"))
+	var extractorUrl fileExtractor = func(r *http.Request) (uploadFile io.Reader, filename string, uerr *UserError) {
+		url := r.FormValue("image")
+		uploadFile, err := s.download(url)
 
 		if err != nil {
-			ErrorResponse(w, "Error dowloading URL!", http.StatusInternalServerError)
-			return
+			return nil, "", &UserError{LogMessage: err, UserFacingMessage: errors.New("Error downloading URL!")}
 		}
 
-		thumbs, err := parseThumbs(r)
-		if err != nil {
-			ErrorResponse(w, err.Error(), http.StatusBadRequest)
-			return
-		}
-
-		s.uploadFile(uploadFile, w, "", thumbs)
-		uploadFile.Close()
+		return uploadFile, path.Base(url), nil
 	}
 
-	base64Handler := func(w http.ResponseWriter, r *http.Request) {
+	var extractorBase64 fileExtractor = func(r *http.Request) (uploadFile io.Reader, filename string, uerr *UserError) {
 		b64data := r.FormValue("image")
 
-		uploadFile := base64.NewDecoder(base64.StdEncoding, strings.NewReader(b64data))
+		uploadFile = base64.NewDecoder(base64.StdEncoding, strings.NewReader(b64data))
 
-		thumbs, err := parseThumbs(r)
-		if err != nil {
-			ErrorResponse(w, err.Error(), http.StatusBadRequest)
-			return
+		return uploadFile, "", nil
+	}
+
+	type uploadEndpoint func(fileExtractor, *AuthenticatedUser) http.HandlerFunc
+
+	var uploadHandler uploadEndpoint = func(extractor fileExtractor, user *AuthenticatedUser) http.HandlerFunc {
+		return func(w http.ResponseWriter, r *http.Request) {
+			uploadFile, filename, uerr := extractor(r)
+			if uerr != nil {
+				log.Printf("Error extracting files: %s", uerr.LogMessage.Error())
+				resp := ServerResponse{
+					Status: http.StatusBadRequest,
+					Error:  uerr.UserFacingMessage.Error(),
+				}
+				resp.Write(w)
+				return
+			}
+
+			thumbs, err := parseThumbs(r)
+			if err != nil {
+				resp := ServerResponse{
+					Status: http.StatusBadRequest,
+					Error:  "Error parsing thumbnails!",
+				}
+				resp.Write(w)
+				return
+			}
+
+			resp := s.uploadFile(uploadFile, filename, thumbs, user)
+
+			switch uploadFile.(type) {
+			case io.ReadCloser:
+				defer uploadFile.(io.ReadCloser).Close()
+				break
+			default:
+				break
+			}
+
+			resp.Write(w)
 		}
+	}
 
-		s.uploadFile(uploadFile, w, "", thumbs)
+	authenticatedEndpoint := func(endpoint uploadEndpoint, extractor fileExtractor) http.HandlerFunc {
```

Comment:
I'm gonna add another test to get coverage on `github.com/Imgur/mandible/server/server.go:88:		NewAuthenticatedServer		0.0%`.